architecture_records/004-dataloader-v2.md

```diff
@@ -0,0 +1,314 @@
+
+1. Implementing data collators as map.
+    We explore this approach as all the processing of data by us is done in data handlers so to implement collators as data handlers 
+    makes sense but is a hard problem to solve. Data collation happens at the very last stage of batch creation and is done in an async
+    fashion because event the HF data collators like `DataCollatorForCompletionOnlyLM` or [`DataCollatorForSeq2Seq`](https://github.com/huggingface/transformers/blob/3f06f95ebe617b192251ef756518690f5bc7ff76/src/transformers/data/data_collator.py#L543) are passed to torch data loader by the HF APIs and is executed by the torch Dataloader.
+
+    So implementing data collators as map means we need to disable [torch data collation and batching](https://pytorch.org/docs/stable/data.html#disable-automatic-batching) and perform it in our code, which we do not intend to experiment with due to its complexity.
+
+1. Passing collators directly to SFTTrainer.
+
+    In the code collators are collected by `get_data_collator` functionality and passed to `SFTTrainer`. We can retain the same functionality
+    and keep the design simpler. The Job of the data loader is to provide a single interface over multiple datasets in the config while keeping a collator like this means we will keep the collator same across all datasets but keeps the design simpler.
+
+## Simplification of code and user configuration
+
+The flexibility provided by this design is that it simplifies the configuration requirement for various use cases.
+If chat template and chat style data is requested users can specify chat specific data handlers and not specify all configurations which are
+not required.
+This can also simplify configuration handling in the code. TODO: give example
+
+## Handling Multi Modal Data.
+
+HF does provide support for handling [image datasets](https://huggingface.co/docs/datasets/en/image_process) and [audio datasets](https://huggingface.co/docs/datasets/en/audio_load) which can be utilised by us in our HF dataloader.
+
+The functionality listed by HF in implementing the use of image and audio datasets is `map` based functions to perform resize, encoding and other such operations on the dataset (see the link above).
+
+This means the image and audio multi modal datasets will be compatible with our data handler routines. Once we have the data handling routines set we will allow users to train with multi modal datasets too.
+
+# Stateful Dataloader implementation.
+
+Stateful dataloader refers to the `fms-fsdp` implementation of the dataloader by our colleagues Davis and Linsong.
+
+Supporting stateful dataloader will mean refactoring the fms-fsdp implementation into our abstract `Dataloader` class.
+
+In brief, things to consider here will be,
+1. Data handler support needs to be added to the stateful data loader as we want lazy execution of handlers (as and when data is loaded).
```

Comment:
its aI think accelerate v1.0 should already have stateful dataloader implemented if I recall correctly?